bindings/matrix-sdk-ffi/src/room_preview.rs

```diff
@@ -1,9 +1,87 @@
-use matrix_sdk::{room_preview::RoomPreview as SdkRoomPreview, RoomState};
+use matrix_sdk::{
+    room_preview::{
+        JoinRoomPreviewAction as SdkJoinRoomPreviewAction, LeaveRoomPreviewAction as SdkLeaveRoomPreviewAction, RoomPreview as SdkRoomPreview,
+        RoomPreviewActions,
+    },
+    Client, RoomState,
+};
 use ruma::space::SpaceRoomJoinRule;
 
-/// The preview of a room, be it invited/joined/left, or not.
+use crate::{client::JoinRule, error::ClientError, room::Membership};
+
 #[derive(uniffi::Record)]
 pub struct RoomPreview {
+    pub info: RoomPreviewInfo,
+    pub room_preview_actions: RoomPreviewActions,
+    pub(crate) sdk_info: matrix_sdk::room_preview::RoomPreview,
+}
+
+#[derive(uniffi::Enum)]
+pub enum RoomPreviewAction {
+    Invited { join: JoinRoomPreviewAction, leave: LeaveRoomPreviewAction },
+    Knocked { leave: LeaveRoomPreviewAction },
+}
+
+#[derive(uniffi::Object)]
+pub struct JoinRoomPreviewAction {
+    inner: SdkJoinRoomPreviewAction,
+}
+
+#[matrix_sdk_ffi_macros::export]
+impl JoinRoomPreviewAction {
+    async fn run(&self) {
+        self.inner.run().unwrap()
+    }
+}
+
+impl From<matrix_sdk::room_preview::RoomPreviewActions> for RoomPreviewAction {
+    fn from(actions: RoomPreviewActions) -> Self {
+        match actions {
+            RoomPreviewActions::Invited { join, leave } => Self::Invited { join, leave },
+            RoomPreviewActions::Knocked { leave } => Self::Knocked { leave },
+        }
+    }
+}
+
+struct
+
+impl RoomPreview {
```

Comment:
Seems like this is not an atomic commit… (It's fine but next time to make the review easier, feel free to squash them altogether, since I know that another commit will now change these three lines.)